### DISCUSSION TITLE
Spring modulith event behavior wrt multiple instance of application

### DISCUSSION THREAD

**Discussion by thereisatechieinme** [spring-projects/spring-modulith#727] [Q&A]
In a Spring Modulith application with multiple instances, how are event publication, processing, and resubmission managed? Specifically, does only one instance listen to the published events, with the first instance to listen handling them? If an instance shuts down abruptly during event processing, will another active instance take over and handle the event reprocessing?

FYI: My application is deployed in AWS via Kubernetes and has 3 application replicas with 1 database instance.

**Comment by odrotbohm** [comment]
As the event distribution is per-application instance, an event in instance A would be processed by listeners within that instance. However, the event publication registry stores the information which listeners need to be invoked in a database. In case of an instance failure, the outstanding event publications can be resubmitted through the `IncompleteEventPublications` abstraction. In a multi-instance setup, you'd have to make sure that this functionality is only performed by a single instance. This in turn is usually implemented using a distributed lock. A popular library providing such functionality is [SchedLock](https://github.com/lukas-krecan/ShedLock).

A lot of this (except the distributed lock approach) is covered in the [reference documentation](https://docs.spring.io/spring-modulith/reference/events.html).

**Reply by raviprakashmishra** [reply]
> As the event distribution is per-application instance, an event in instance A would be processed by listeners within that instance. However, the event publication registry stores the information which listeners need to be invoked in a database. In case of an instance failure, the outstanding event publications can be resubmitted through the `IncompleteEventPublications` abstraction. In a multi-instance setup, you'd have to make sure that this functionality is only performed by a single instance. This in turn is usually implemented using a distributed lock. A popular library providing such functionality is [SchedLock](https://github.com/lukas-krecan/ShedLock).
> 
> A lot of this (except the distributed lock approach) is covered in the [reference documentation](https://docs.spring.io/spring-modulith/reference/events.html).

Only issue in this approach is that there could be a race condition where your Shedlock job picks up an event which is already being processed by the local/instance specific event bus.  This may not be a problem in 2.0 considering we have various event states.

**Comment by tomdevroomen** [comment]
@odrotbohm Also for Spring Modulith 2.0 a multi-instance setup isn't covered, right?
To have a guaranteed order of republishing failed events, you'll still need some distributed locking mechanism, to be sure only 1 instance is handling this.

By default Spring Modulith supports resubmission on restart only.
If there's an desire to have some retry mechanism, some homebrewed scheduling is required to have resubmission on a running application?

**Comment by mf-guse** [comment]
It would be nice to have eventhandling which is useable in a multi-instance setup.

In our case we have multiple instances of our service and we would like to ensure that an async (and persisted) event is processed only once in the multi-instance setup and it does not matter which instance does the processing.

Is there any working example which shows how to achieve this using spring-modulith 2?
